thingsboard_gateway/gateway/tb_gateway_service.py

```diff
@@ -634,23 +634,25 @@ def __send_to_storage(self):
 
                         max_data_size = self.__config["thingsboard"].get("maxPayloadSizeBytes", 1024)
                         if self.__get_data_size(data) >= max_data_size:
+                            # Data is too large, so we will attempt to send in pieces
                             adopted_data = {"deviceName": data['deviceName'],
                                             "deviceType": data['deviceType'],
                                             "attributes": {},
                                             "telemetry": []}
+
+                            # First, loop through the attributes
                             for attribute in data['attributes']:
+                                adopted_data['attributes'].update({attribute: data['attributes'][attribute]})
                                 adopted_data_size = self.__get_data_size(adopted_data)
                                 if adopted_data_size >= max_data_size:
+                                    # We have surpassed the max_data_size, so send what we have and clear attributes
                                     self.__send_data_pack_to_storage(adopted_data, connector_name)
                                     adopted_data['attributes'] = {}
-                                adopted_data['attributes'].update({attribute: data['attributes'][attribute]})
+                            
+                            # Now, loop through telemetry. Possibly have some unsent attributes that have been adopted.
                             for ts_kv_list in data['telemetry']:
```

Comment:
Before this line it has to be the next line:
`telemetry = data['telemetry'] if isinstance(data['telemetry'], list) else [data['telemetry']]`

Because see on method `__convert_telemetry_to_ts` - it can return dict and list (so if it is dict we have the exception in the below lines where we want to get ts from current iter dict.